File: c/sorting/standard/merge_sort.c
#include <stdlib.h>

int* merge_sort(int *arr, int n)
{
    if (n <= 1)
    {
        int *result = malloc(sizeof(int) * n);
        for (int i = 0; i < n; i++) result[i] = arr[i];
        return result;
    }

    int mid = n / 2;
    int *left = merge_sort(arr, mid);
    int *right = merge_sort(arr + mid, n - mid);
    int *result = malloc(sizeof(int) * n);
    int i = 0, j = 0, k = 0;

    while (i < mid && j < n - mid)
        result[k++] = left[i] <= right[j] ? left[i++] : right[j++];
    while (i < mid) result[k++] = left[i++];
    while (j < n - mid) result[k++] = right[j++];

    free(left);
    free(right);
    return result;
}

File: java/sorting/standard/MergeSort.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

public class MergeSort {

    public static int[] mergeSort(int[] arr) {
        if (arr.length <= 1) {
            return Arrays.copyOf(arr, arr.length);
        }

        int mid = arr.length / 2;
        int[] left = mergeSort(Arrays.copyOfRange(arr, 0, mid));
        int[] right = mergeSort(Arrays.copyOfRange(arr, mid, arr.length));

        List<Integer> result = new ArrayList<>();
        int i = 0;
        int j = 0;

        while (i < left.length && j < right.length) {
            if (left[i] <= right[j]) {
                result.add(left[i++]);
            } else {
                result.add(right[j++]);
            }
        }

        while (i < left.length) {
            result.add(left[i++]);
        }
        while (j < right.length) {
            result.add(right[j++]);
        }

        return result.stream().mapToInt(x -> x).toArray();
    }
}
